// vike/dist/cjs/node/plugin/plugins/setGlobalContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setGlobalContext = setGlobalContext;
const globalContext_js_1 = require("../../runtime/globalContext.js");
const utils_js_1 = require("../utils.js");
const getVikeConfig_js_1 = require("./importUserCode/v1-design/getVikeConfig.js");
const getViteConfigRuntime_js_1 = require("../shared/getViteConfigRuntime.js");
function setGlobalContext() {
    let isServerReload = false;
    let config;
    return [
        {
            name: 'vike:setGlobalContext:pre',
            enforce: 'pre',
            // This hook is called not only at server start but also at server restart (a new `viteDevServer` instance is created)
            configureServer: {
                order: 'pre',
                handler(viteDevServer) {
                    (0, utils_js_1.assert)(config);
                    if (isServerReload)
                        (0, getVikeConfig_js_1.reloadVikeConfig)(config);
                    isServerReload = true;
                    (0, globalContext_js_1.setGlobalContext_viteDevServer)(viteDevServer);
                    (0, utils_js_1.markSetup_viteDevServer)();
                }
            },
            configurePreviewServer() {
                (0, utils_js_1.markSetup_vitePreviewServer)();
            },
            config: {
                order: 'pre',
                handler(_, env) {
                    const isViteDev = (0, utils_js_1.isDevCheck)(env);
                    (0, globalContext_js_1.setGlobalContext_isProduction)(!isViteDev);
                    (0, utils_js_1.markSetup_isViteDev)(isViteDev);
                }
            }
        },
        {
            name: 'vike:setGlobalContext:post',
            enforce: 'post',
            configResolved: {
                order: 'post',
                async handler(config_) {
                    config = config_;
                    const viteConfigRuntime = (0, getViteConfigRuntime_js_1.getViteConfigRuntime)(config);
                    (0, globalContext_js_1.setGlobalContext_viteConfig)(config, viteConfigRuntime);
                }
            }
        }
    ];
}
