// @webstudio-is/sdk-components-react-radix/lib/dialog.ws.js
import { DialogIcon as e, TriggerIcon as n, OverlayIcon as i, ContentIcon as r, ButtonElementIcon as a, HeadingIcon as s, TextIcon as l } from "@webstudio-is/icons/svg";
import { defaultStates as c } from "@webstudio-is/sdk";
import { div as t, button as p, h2 as g, p as D } from "@webstudio-is/sdk/normalize.css";
import { radix as o } from "./shared/meta.js";
import { propsDialog as d, propsDialogTrigger as m, propsDialogOverlay as y, propsDialogContent as M, propsDialogClose as h, propsDialogTitle as C, propsDialogDescription as T } from "./__generated__/dialog.props.js";
import { buttonReset as f } from "./shared/preset-styles.js";
const O = {
  icon: n,
  contentModel: {
    category: "none",
    children: ["instance"]
  }
}, b = {
  icon: i,
  contentModel: {
    category: "none",
    children: ["instance"],
    descendants: [o.DialogContent]
  },
  presetStyle: {
    div: t
  }
}, B = {
  icon: r,
  contentModel: {
    category: "none",
    children: ["instance"],
    descendants: [
      o.DialogTitle,
      o.DialogDescription,
      o.DialogClose
    ]
  },
  presetStyle: {
    div: t
  }
}, E = {
  icon: s,
  contentModel: {
    category: "none",
    children: ["instance", "rich-text"]
  },
  presetStyle: {
    h2: g
  }
}, H = {
  icon: l,
  contentModel: {
    category: "none",
    children: ["instance", "rich-text"]
  },
  presetStyle: {
    p: D
  }
}, R = {
  icon: a,
  label: "Close Button",
  contentModel: {
    category: "none",
    children: ["instance", "rich-text"]
  },
  states: c,
  presetStyle: {
    button: [f, p].flat()
  }
}, j = {
  icon: e,
  contentModel: {
    category: "instance",
    children: ["instance"],
    descendants: [o.DialogTrigger, o.DialogOverlay]
  }
}, k = {
  props: d,
  initialProps: []
}, q = {
  props: m
}, w = {
  props: M,
  initialProps: []
}, z = {
  props: y,
  initialProps: []
}, A = {
  props: h,
  initialProps: []
}, F = {
  props: C,
  initialProps: []
}, G = {
  props: T,
  initialProps: []
};
export {
  j as metaDialog,
  R as metaDialogClose,
  B as metaDialogContent,
  H as metaDialogDescription,
  b as metaDialogOverlay,
  E as metaDialogTitle,
  O as metaDialogTrigger,
  k as propsMetaDialog,
  A as propsMetaDialogClose,
  w as propsMetaDialogContent,
  G as propsMetaDialogDescription,
  z as propsMetaDialogOverlay,
  F as propsMetaDialogTitle,
  q as propsMetaDialogTrigger
};
