// @webstudio-is/sdk-components-react-radix/lib/tooltip.ws.js
import { TooltipIcon as t, TriggerIcon as n, ContentIcon as e } from "@webstudio-is/icons/svg";
import { div as i } from "@webstudio-is/sdk/normalize.css";
import { radix as o } from "./shared/meta.js";
import { propsTooltip as r, propsTooltipTrigger as p, propsTooltipContent as s } from "./__generated__/tooltip.props.js";
const g = {
  icon: n,
  contentModel: {
    category: "none",
    children: ["instance"]
  }
}, d = {
  icon: e,
  contentModel: {
    category: "none",
    children: ["instance"]
  },
  presetStyle: {
    div: i
  }
}, m = {
  icon: t,
  contentModel: {
    category: "instance",
    children: ["instance"],
    descendants: [o.TooltipTrigger, o.TooltipContent]
  }
}, f = {
  props: r,
  initialProps: ["open", "delayDuration", "disableHoverableContent"]
}, C = {
  props: p
}, M = {
  props: s,
  initialProps: ["side", "sideOffset", "align", "alignOffset"]
};
export {
  m as metaTooltip,
  d as metaTooltipContent,
  g as metaTooltipTrigger,
  f as propsMetaTooltip,
  M as propsMetaTooltipContent,
  C as propsMetaTooltipTrigger
};
