// @webstudio-is/sdk-components-react/lib/box.ws.js
import { BoxIcon as t } from "@webstudio-is/icons/svg";
import { defaultStates as e } from "@webstudio-is/sdk";
import { section as r, nav as o, main as a, header as i, footer as s, figure as n, aside as d, article as p, address as c, div as m } from "@webstudio-is/sdk/normalize.css";
import { props as f } from "./__generated__/box.props.js";
const v = {
  category: "general",
  description: "A container for content. By default this is a Div, but the tag can be changed in settings.",
  icon: t,
  states: e,
  presetStyle: {
    div: m,
    address: c,
    article: p,
    aside: d,
    figure: n,
    footer: s,
    header: i,
    main: a,
    nav: o,
    section: r
  },
  order: 0
}, y = {
  props: {
    ...f,
    tag: {
      required: !0,
      control: "tag",
      type: "string",
      options: [
        "div",
        "header",
        "footer",
        "nav",
        "main",
        "section",
        "article",
        "aside",
        "address",
        "figure",
        "span"
      ]
    }
  },
  initialProps: ["tag", "id", "className"]
};
export {
  v as meta,
  y as propsMeta
};
